lib/src/main/java/io/ably/lib/types/MessageExtras.java

```diff
@@ -1,64 +1,150 @@
 package io.ably.lib.types;
 
 import com.google.gson.JsonElement;
-import io.ably.lib.util.Log;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
 import io.ably.lib.util.Serialisation;
-import org.msgpack.core.MessageFormat;
 import org.msgpack.core.MessagePacker;
 import org.msgpack.core.MessageUnpacker;
+import org.msgpack.value.ImmutableMapValue;
+import org.msgpack.value.ImmutableValue;
+import org.msgpack.value.Value;
+import org.msgpack.value.ValueFactory;
 
 import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.Objects;
 
 public final class MessageExtras {
 	private static final String TAG = MessageExtras.class.getName();
-	
-	// TODO what about the 1.1 push extension? (TM2i)
 
-	private final DeltaExtras delta; // may be null
-	
+	private static final String DELTA = "delta";
+
+	private final DeltaExtras delta;
+	private final JsonObject raw;
+
+	/**
+	 * Creates a MessageExtras instance to be sent as extra with a Message to Ably's servers.
+	 *
+	 * @see <a href="https://www.ably.io/documentation/general/push/publish#channel-broadcast-example">Channel-based push notification example</a>
+	 *
+	 * @since 1.2.1
+	 */
+	public MessageExtras(final JsonObject raw) {
+		this.raw = raw;
+		delta = null;
+	}
+
+	/**
+	 * @since 1.2.0
+	 */
 	public MessageExtras(final DeltaExtras delta) {
 		this.delta = delta;
+		raw = null;
 	}
-	
+
+	private MessageExtras(final JsonObject raw, final DeltaExtras delta) {
+		this.raw = raw;
+		this.delta = delta;
+	}
+
 	public DeltaExtras getDelta() {
 		return delta;
 	}
 
-	/* package private */ JsonElement toJsonElement() {
-		return Serialisation.gson.toJsonTree(this);
+	/* package private */ JsonObject getRaw() {
+		return raw;
```

Comment:
Doesn't this need to be `public` ? If someone validly sends me a message containing a member in `extras` that this library doesn't know about, then I'm going to have to process the raw JSON.

If we are going to make it public, then perhaps it needs to be `asJSON()` or something like that.

Reply:
Happy to make that public, yes. I was just trying to minimise exposure but I can see that's a clearly valid use case. I agree that, in that case, the name would need to change too.

Reply:
I've now made that method public: https://github.com/ably/ably-java/pull/581/commits/e41c806d707f77e6755d19c1eb8555aa0688ef0a